cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/SparkRuntimeUtils.java

```diff
@@ -232,10 +258,68 @@ private void handleCompleted(boolean failure) {
         if (failure && driverService instanceof SparkDriverService) {
           ((SparkDriverService) driverService).stopWithoutComplete();
         } else {
-          driverService.stopAndWait();
+          driverService.stop();
         }
+
+        // Wait for the spark execution context to stop to make sure everything related
+        // to the spark execution is shutdown properly before returning.
+        // When this method returns, the driver's main thread returns, which will have system services shutdown
+        // and process terminated. If system services stopped concurrently with the execution context, we may have some
+        // logs and tasks missing.
+        Uninterruptibles.awaitUninterruptibly(secStopLatch, 30, TimeUnit.SECONDS);
```

Comment:
Is it possible to log a warning if the wait times out? That way if logs are missing we can easily tell that is was due to the timeout or some other reason.

Reply:
Added a log